vSAN methods QueryVsanObjects and VsanQueryObjectIdentities has been added

marunachalam opened the pull request:
Added couple of vsan api interface which is required for my e2e-test automation

Testing Done:
go test -v

```
=== RUN   TestClient
--- PASS: TestClient (4.04s)
=== RUN   TestVsanQueryObjectIdentities
    TestVsanQueryObjectIdentities: client_test.go:128: Printing clusterConfig:
         &types.VsanObjectIdentityAndHealth{
            Health:       (*types.VsanObjectOverallHealth)(nil),
            Identities:   &types.VsanObjectIdentity{Uuid:"84c65a5f-46ed-c821-dedf-02005e6ac5d0"},
            RawData:      "",
            SpaceSummary: nil,
        }
--- PASS: TestVsanQueryObjectIdentities (4.71s)
PASS
ok      github.com/vmware/govmomi/vsan  8.770s
```

gh05tn0va commented on vsan/client.go:
```diff
@@ -84,3 +88,95 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+
+// Creates the vsan object identities instance. This is to be queried from vsan health.
+var (
+	VsanQueryObjectIdentitiesInstance = vimtypes.ManagedObjectReference{
```
VsanObjectSystemInstance

gh05tn0va commented on vsan/client.go:
```diff
@@ -84,3 +88,95 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+
+// Creates the vsan object system instance. This is to be queried from vsan health.
+var (
+	QueryVsanObjectsInstance = vimtypes.ManagedObjectReference{
```
HostVsanInternalSystemInstance. And this is not used, right?

gh05tn0va commented on vsan/client.go:
```diff
@@ -84,3 +88,95 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+	finder.SetDatacenter(dc)
+
+	clusterComputeResource, err := finder.ClusterComputeResourceList(ctx, "*")
+	cluster := clusterComputeResource[0]
```
This will get you host from the first cluster. What if we want to query from host in another cluster?

marunachalam replied:
Have removed this block, since its not required

divyenpatel commented:
@marunachalam Can you please squash all your commits.

I see you have not updated `vsan/client.go`. How are you planning to call these methods?

Can you share testing done for these methods? Can you implement some tests and add it to `vsan/client_test.go`.

marunachalam commented:
> @marunachalam Can you please squash all your commits.
> 
> I see you have not updated `vsan/client.go`. How are you planning to call these methods?
> 
> Can you share testing done for these methods? Can you implement some tests and add it to `vsan/client_test.go`.

I have added the client and test methods

Test Result -

=== RUN   TestClient
--- PASS: TestClient (4.94s)
=== RUN   TestVsanQueryObjectIdentities
    TestVsanQueryObjectIdentities: client_test.go:128: Printing clusterConfig:
         &types.VsanObjectIdentityAndHealth{
            Health:       (*types.VsanObjectOverallHealth)(nil),
            Identities:   &types.VsanObjectIdentity{Uuid:"b004315f-d6d8-771f-8ef9-020055b45c1b"},
            RawData:      "",
            SpaceSummary: nil,
        }
--- PASS: TestVsanQueryObjectIdentities (6.29s)
=== RUN   TestQueryVsanObjects
    TestQueryVsanObjects: client_test.go:183: No vSAN objects found
--- PASS: TestQueryVsanObjects (5.36s)
PASS
ok      github.com/vmware/govmomi/vsan  16.622s

divyenpatel commented:
@marunachalam Thanks for addressing all review comments. Can you please squash all commits in this PR?

marunachalam commented:
> @marunachalam Thanks for addressing all review comments. Can you please squash all commits in this PR?

Thanks for looking into it @divyenpatel , I have squashed my commits.

divyenpatel commented:
/lgtm

divyenpatel commented:
@dougm Can you help review this PR?

dougm commented on vsan/client.go:
```diff
@@ -84,3 +88,72 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+// QueryVsanObjects Queries the vsan object
+func (c *Client) QueryVsanObjects(ctx context.Context, uuids []string) (string, error) {
+
+	url := os.Getenv("VC_URL") // example: export VC_URL='https://username:password@vc-ip/sdk'
```
I don't think we want to use env vars with the `Client` here, which already has a URL configured you can access using `c.vim25Client.URL()`

marunachalam replied:
Sure, Now I have passed it as param

dougm commented on vsan/client.go:
```diff
@@ -84,3 +88,72 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+func (c *Client) QueryVsanObjects(ctx context.Context, uuids []string) (string, error) {
+
+	url := os.Getenv("VC_URL") // example: export VC_URL='https://username:password@vc-ip/sdk'
+	datacenter := os.Getenv("DATACENTER")
```
Rather than env var here, this should probably be a `types.ManagedObjectReference` param to the method.

marunachalam replied:
Same as above, I have passed the it as a param

dougm commented on vsan/client.go:
```diff
@@ -84,3 +88,72 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+	dc, err := finder.Datacenter(ctx, datacenter)
+	finder.SetDatacenter(dc)
+
+	clusterComputeResource, err := finder.ClusterComputeResourceList(ctx, "*")
```
Need to check `err != nil` here

marunachalam replied:
done

dougm commented on vsan/client.go:
```diff
@@ -84,3 +88,72 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+	finder.SetDatacenter(dc)
+
+	clusterComputeResource, err := finder.ClusterComputeResourceList(ctx, "*")
+	cluster := clusterComputeResource[0]
```
Check `len(clusterComputeResource) != 0` ?

marunachalam replied:
I have added a check for the same

dougm commented on vsan/client.go:
```diff
@@ -84,3 +88,72 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+	clusterComputeResource, err := finder.ClusterComputeResourceList(ctx, "*")
+	cluster := clusterComputeResource[0]
+	config, err := cluster.Configuration(ctx)
+	hostSystem := config.VsanHostConfig[0].HostSystem.String()
```
It looks like this could be any host in any cluster, is that the case?

marunachalam replied:
yes @dougm

divyenpatel commented on vsan/client.go:
```diff
@@ -84,3 +87,78 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+
+	u, err := soap.ParseURL(url)
+	ctx = context.Background()
+	c1, err := govmomi.NewClient(ctx, u, true)
```
why are we creating a new client?
function should use the `client` already defined in the func definition. `func (c *Client) QueryVsanObjects`.

divyenpatel commented on vsan/client.go:
```diff
@@ -84,3 +87,78 @@ func (c *Client) VsanPerfQueryPerf(ctx context.Context, cluster *vimtypes.Manage
+}
+
+// QueryVsanObjects Queries the vsan object
+func (c *Client) QueryVsanObjects(ctx context.Context, uuids []string, datacenter string, url string) (string, error) {
```
`QueryVsanObjects` should not expect the URL. URL will have a username and password.
this function should utilize the client.

divyenpatel commented:
@marunachalam change looks good to me. Can you please squash all your commit?

divyenpatel commented:
@marunachalam Can you add a test execution log in the PR description?

marunachalam commented:
Testing Done:
go test -v
```
=== RUN   TestClient
--- PASS: TestClient (4.04s)
=== RUN   TestVsanQueryObjectIdentities
    TestVsanQueryObjectIdentities: client_test.go:128: Printing clusterConfig:
         &types.VsanObjectIdentityAndHealth{
            Health:       (*types.VsanObjectOverallHealth)(nil),
            Identities:   &types.VsanObjectIdentity{Uuid:"84c65a5f-46ed-c821-dedf-02005e6ac5d0"},
            RawData:      "",
            SpaceSummary: nil,
        }
--- PASS: TestVsanQueryObjectIdentities (4.71s)
PASS
ok      github.com/vmware/govmomi/vsan  8.770s
```

marunachalam commented:
> @marunachalam change looks good to me. Can you please squash all your commit?

I have squashed the commits. Thanks Divyen.

marunachalam commented:
> @marunachalam Can you add a test execution log in the PR description?

Done, Added the same

dougm reviewed (approved):
Thanks @marunachalam
